index/scorch/persister.go

```diff
@@ -871,24 +886,79 @@ func (s *Scorch) removeOldData() {
 // rollback'ability.
 var NumSnapshotsToKeep = 1
 
+// RollbackSamplingInterval controls how far back we are looking
+// in the history to get the rollback points.
+// For example, the default value of 1 hour ensures that the
+// protected snapshots (NumSnapshotsToKeep = 3) are:
+//    the very latest snapshot(ie the current one),
+//    the snapshot that was persisted 1 hour before the current one,
+//	  the snapshot that was persisted 2 hours before the current one
+var RollbackSamplingInterval = 1 * time.Hour
+
+// getProtectedEpochs aims to fetch the epochs keep based on a timestamp basis.
+// It tries to get NumSnapshotsToKeep snapshots, each of which are separated
+// by a time duration of RollbackSamplingInterval.
+func (s *Scorch) getProtectedEpochs(
+	persistedSnapshots []*snapshotMetaData) map[uint64]struct{} {
+
+	// make a map of epochs to protect from deletion
+	protectedEpochs := make(map[uint64]struct{}, s.numSnapshotsToKeep)
+	protectedEpochs[persistedSnapshots[0].epoch] = struct{}{}
+	nextSnapshotToProtect :=
+		persistedSnapshots[0].timeStamp.Add(s.rollbackSamplingInterval)
+	protectedSnapshots := 1
+	lastProtectedSnapshot := 0
+
+	for i := 1; i < len(persistedSnapshots); i++ {
+		// Finding that snapshot whose timestamp is just above
+		// the required timestamp, because the snapshot at the previous
+		// index is bound to be the closest possible snapshot to the
+		// required timestamp of nextSnapshotToProtect
+		if persistedSnapshots[i].timeStamp.After(nextSnapshotToProtect) {
+			protectedEpochs[persistedSnapshots[i-1].epoch] = struct{}{}
```

Comment:
Hmm this doesn't feel right, shouldn't you be adding to the map ..
```
protectedEpochs[persistedSnapshots[i].epoch] = struct{}{}
```

Reply:
nextSnapshotToProtect essentially indicates the timestamp of the next snapshot we *want* to protect. There can be a situation that we don't have a snapshot in our list of persisted snapshots, with the exact timestamp that we're trying to find, during which we just protect the snapshot closest to the required timestamp, but less than it. So, we are trying to find the first timestamp (at index `i`) that's greater than the nextSnapshotToProtect and we can say for sure that the snapshot at `i-1` has a timestamp lesser than nextSnapshotToProtect which is bound to be the closest one.